### DISCUSSION TITLE
Pyodide with SDL2 support.

### DISCUSSION THREAD

**Discussion by km19809** [pyodide/pyodide#2986] [Show and tell]
Hello.
I want to show a (WIP) work that uses pyodide and SDL2.
Here is one of the [demo](https://km19809.github.io/pyxelodide/test_pyxel_fetch.html?py=00_moving_square)

It runs [pyxel](https://github.com/kitao/pyxel), which is one of the greatest python 2D game engines.
Although the porting was not finished yet, I want to share the result here.
Without the support from the contributors of pyodide, I was not able to port them. (Of course, thanks to participants of https://github.com/kitao/pyxel/issues/354 !)

Its main differences from the original are:
1. It was Linked with SDL2. (https://github.com/pyodide/pyodide/issues/2964#issuecomment-1214731392)
2. It detects `canvas` automatically, by modifying [`pre.js`](https://emscripten.org/docs/tools_reference/emcc.html#emcc-pre-js)
3. It uses the pre-released version of SDL2, to solve the audio buffer problem. (https://github.com/libsdl-org/SDL/issues/6060)

Of course, there are some weird behaviors (especially related to the mouse).
Also, other SDL2 packages are not tested (ex: Pygame, PySDL2; They need additional libs such as SDL2_TTF).
But I'm glad that I can share my experience.

It will be great if pyodide support SDL2 officially. 
(At least, provide a version with SDL2 support. It makes compiled binary larger, so it loads slow.)

Here's my fork:
[github](https://github.com/km19809/km19809.github.io/tree/main/pyxelodide)

**Comment by ryanking13** [comment]
Thanks for sharing your work @km19809. Are you using SDL2 port that emscripten provides or did you build SDL2 by userself? I think the emscripten version of SDL2  was a bit old.

> It will be great if pyodide support SDL2 officially.

Should SDL2 need to be linked to the Pyodide main module?

**Reply by km19809** [reply]
We can change the version by modifying some variables. (https://github.com/emscripten-core/emscripten/discussions/17652)
> Should SDL2 need to be linked to the Pyodide main module?

Yes, without it, the emscripten does not generate SDL-related code. (https://github.com/pyodide/pyodide/issues/2964#issuecomment-1214651891)

**Reply by ryanking13** [reply]
> Yes, without it, the emscripten does not generate SDL-related code. ([#2964 (comment)](https://github.com/pyodide/pyodide/issues/2964#issuecomment-1214651891))

> because only main module has system libraries linked in

I think Pyodide is working pretty well with linking libraries to side modules, but, I don't know, maybe SDL2 can be an exception.
How does the size of `pyodide.asm.wasm` changes when you link SDL2 to the main module?

**Reply by km19809** [reply]
Before:
<img width="389" alt="image" src="https://user-images.githubusercontent.com/20869932/185344481-7a19483a-648d-4ef2-b1f7-365b3f6f5370.png">

wasm 3.0MB, data 3.2MB, asm.js 278kB
After:
<img width="389" alt="image" src="https://user-images.githubusercontent.com/20869932/185344362-0742723c-add0-44d5-8ba8-1d37db671d5f.png">
wasm 9.6MB, data 5.1MB, asm.js 1.8MB

At least 10MB larger. I think there is something that I missed (especially optimization).

**Reply by ryanking13** [reply]
In main, the size of `pyodide.asm.wasm` is around 8MB, maybe you detached some libraries from the main module?

**Reply by ryanking13** [reply]
> > because only main module has system libraries linked in
> 
> I think Pyodide is working pretty well with linking libraries to side modules

I suspect https://github.com/emscripten-core/emscripten/issues/17675 this was the reason why SDL2 was not working correctly when linked to the side modules.

**Reply by hoodmane** [reply]
I think there is also the more fundamental problem that SDL2 includes JavaScript code, and the dynamic linking model does not currently support JavaScript in side modules.
